# 1. Player value records
In this notebook we explore the available player value records. They are stored in a dataframe in pickle format.

In [1]:
import pandas as pd
values = pd.read_pickle('../data/value_records_for_ratings_based_predictions.pkl')

In [2]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import linear_model 

In [3]:
values.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9610 entries, 5665 to 160599
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   player_id    9610 non-null   int32         
 1   player_name  9610 non-null   object        
 2   player_role  9607 non-null   category      
 3   birth        9610 non-null   datetime64[ns]
 4   height       9610 non-null   float64       
 5   foot         9610 non-null   object        
 6   value        9610 non-null   float64       
 7   league       9610 non-null   object        
 8   value_at     9610 non-null   datetime64[ns]
 9   nat1         9610 non-null   object        
 10  nat2         9610 non-null   object        
dtypes: category(1), datetime64[ns](2), float64(2), int32(1), object(5)
memory usage: 797.8+ KB


Here is a little explanation of the data contained in the `values` data frame.
- `player_id` is the TransferMarkt id of the player
- `player_name`, `player_role`, `height`, `foot`, and `birt` are self-explanatory

In [4]:
values['birth'].describe()

/var/folders/tw/m4_1qbzn2ybbmcwkchr6k6_c0000gp/T/ipykernel_2774/3604468148.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  values['birth'].describe()


count                    9610
unique                   2616
top       1994-05-27 00:00:00
freq                       22
first     1977-01-02 00:00:00
last      2001-08-16 00:00:00
Name: birth, dtype: object

- `nat1` and `nat2` are string-valued columns that store the first and second nationality of the players which have more than one.

In [5]:
values.loc[:,['player_name','nat1','nat2']].head()

player_name       nat1       nat2
5665    Marko Dmitrovic     Serbia          -
5667     Paulo Oliveira   Portugal          -
5671       José Ángel      Spain          -
5674  Gonzalo Escalante  Argentina      Italy
5676              Bebé   Portugal  CapeVerde

- `value` contains the market value of the player recorded at time `value_at`.
- `league` is the league in which the player was competing at time `value_at`. 

In [6]:
values.loc[:,['player_id','player_name','birth','value','value_at','league']].head()

player_id        player_name      birth  value   value_at league
5665      94308    Marko Dmitrovic 1992-01-24   3.60 2018-06-30   SPA1
5667     139336     Paulo Oliveira 1992-01-08   2.70 2018-06-30   SPA1
5671      87469       José Ángel 1989-09-05   2.25 2018-06-30   SPA1
5674     266795  Gonzalo Escalante 1993-03-27   2.25 2018-06-30   SPA1
5676     153427              Bebé 1990-07-12   0.90 2018-06-30   SPA1

We can, for example, extract all records for the players born in a certain time interval (e.g., born in 2000).

In [7]:
values.loc[(values['birth']> '2000-1-1') & (values['birth'] < '2000-12-31'),:].head()

player_id    player_name         player_role      birth  height   foot  \
18850     316760  Vincent Thill  Attacking Midfield 2000-02-04    1.70   left   
19488     370846   Timothy Weah      Centre-Forward 2000-02-22    1.85  right   
21177     464990    Arton Zekaj  Defensive Midfield 2000-04-16    1.87  right   
23745     463665    Oumar Solet         Centre-Back 2000-02-07    1.92  right   
23758     418659   Amine Gouiri      Centre-Forward 2000-02-16    1.80  right   

       value league   value_at          nat1                    nat2  
18850  0.540    FR1 2017-06-30    Luxembourg                       -  
19488  0.900    FR1 2018-06-30  UnitedStates                  France  
21177  0.045    FR1 2018-06-30        Kosovo                  Serbia  
23745  0.900    FR1 2018-06-30        France  CentralAfricanRepublic  
23758  4.050    FR1 2018-06-30        France                 Algeria

We can further narrow down the search and select the records for the players born in a certain year (e.g., 1990) when they where 25 years old, i.e., in 2015.

In [8]:
values.loc[(values['birth']> '1990-1-1') & (values['birth'] < '1990-12-31') & (values['value_at'] == '2015-06-30') ,:]

player_id          player_name         player_role      birth  height  \
9768       133794        Edgar Méndez        Right Winger 1990-01-02    1.87   
9779        73092           Rene Krhin  Defensive Midfield 1990-05-21    1.89   
10313      203043          Pedro Bigas         Centre-Back 1990-05-15    1.81   
10321      223068                 Tana  Attacking Midfield 1990-09-20    1.69   
13754       59344   Asier Illarramendi  Defensive Midfield 1990-03-08    1.79   
...           ...                  ...                 ...        ...     ...   
158151      45184  Grzegorz Krychowiak  Defensive Midfield 1990-01-29    1.87   
158621     183647    Álvaro González         Centre-Back 1990-01-08    1.82   
159548      58426              Douglas          Right-Back 1990-08-06    1.72   
160482      58884                Nacho         Centre-Back 1990-01-18    1.80   
160487      31909           Toni Kroos    Central Midfield 1990-01-04    1.83   

         foot  value league   value_at      nat1 nat2  
9768    right   2.25   SPA1 2015-06-30     Spain    -  
9779    right   2.25   SPA1 2015-06-30  Slovenia    -  
10313    left   0.90   SPA1 2015-06-30     Spain    -  
10321    left   0.18   SPA1 2015-06-30     Spain    -  
13754   right  13.50   SPA1 2015-06-30     Spain    -  
...       ...    ...    ...        ...       ...  ...  
158151  right  10.80   SPA1 2015-06-30    Poland    -  
158621  right   2.70   SPA1 2015-06-30     Spain    -  
159548  right   1.80   SPA1 2015-06-30    Brazil    -  
160482  right   5.40   SPA1 2015-06-30     Spain    -  
160487   both  45.00   SPA1 2015-06-30   Germany    -  

[115 rows x 11 columns]

Let us simplify a bit players roles using macro roles

In [9]:
list(values['player_role'].unique())

['Goalkeeper',
 'Centre-Back',
 'Left-Back',
 'Central Midfield',
 'Left Winger',
 'Right Winger',
 'Centre-Forward',
 'Right-Back',
 'Defensive Midfield',
 'Attacking Midfield',
 'Second Striker',
 'Right Midfield',
 'Left Midfield',
 nan]

In [10]:
macro_role = {'Goalkeeper':'GK', 
              'Centre-Back':'DF',
              'Defensive Midfield': 'MF',
              'Right Winger':'MF',
              'Centre-Forward':'FW',
              'Right-Back':'DF',
              'Attacking Midfield':'MF',
              'Central Midfield':'MF',
              'Left-Back':'DF',
              'Left Winger':'MF',
              'Right Midfield':'MF',
              'Left Midfield':'MF',
              'Second Striker':'FW',
              'Attacking Midfield':'MF'
             }

In [11]:
values["macro_role"] = values["player_role"].apply(lambda x: macro_role.get(x))
values.head()

player_id        player_name       player_role      birth  height  \
5665      94308    Marko Dmitrovic        Goalkeeper 1992-01-24    1.94   
5667     139336     Paulo Oliveira       Centre-Back 1992-01-08    1.87   
5671      87469       José Ángel         Left-Back 1989-09-05    1.82   
5674     266795  Gonzalo Escalante  Central Midfield 1993-03-27    1.82   
5676     153427              Bebé       Left Winger 1990-07-12    1.90   

       foot  value league   value_at       nat1       nat2 macro_role  
5665   left   3.60   SPA1 2018-06-30     Serbia          -         GK  
5667  right   2.70   SPA1 2018-06-30   Portugal          -         DF  
5671   left   2.25   SPA1 2018-06-30      Spain          -         DF  
5674  right   2.25   SPA1 2018-06-30  Argentina      Italy         MF  
5676  right   0.90   SPA1 2018-06-30   Portugal  CapeVerde         MF

# 2. Filtering out missing players
For a number of players in the values records there are no ratings statistics. This might be due to the fact that they have not made an appearence in official matches. We filter them out.

In [12]:
missing = pd.read_csv('../data/missing.csv')
len(missing)

41

In [13]:
missing.head()

ID                Name
0  250038     Mamadou Sissako
1  361082  Jasper Schendelaar
2  215094           Nick Olij
3  387241      Raman ten Hove
4  153751      Nick Hengelman

In [14]:
len(values)

9610

In [15]:
values = values.loc[~values['player_id'].isin(missing['ID']), :]
len(values) 

9556

This number is smaller than 9610 - 41 = 9559 as some players may be in more records, i.e., have value for different years.

# 3. Ranking
Market values are not predicted directly, that is, not the dollar/euro value of the player. This is due to the fact that this value may be influenced by several external factors such as the general economic situation. Thus, we cannot e.g., compare a dollar value of 2010 and a dollar value of 2020, even if discounted. Instead, we predict the *ranking* of the player in a value table listing all players born in the same year and having the same age. 
In fact, we cannot simply rank all players at the age of, say, 25. The value of a 25 years old player in e.g., 2010 cannot be compared directly with the value of a 25 years old player in 2015. Rather, we should divide players based on their birth year and age. As an example, we could rank all the 25 years old players born in 1990.
This is done as follows
- We divide the players according to their birth year. Thus we will have a set of players born in 1990, a set of players born in 1991 and so on.
- For each birth year, we divide the players based on age, thus, for the players born in, say, 1990, we we divide the records with `value_at` in year 2010 (age 20), in year 2011 (age 21) and so on.
- We rank the players in each birth-age group in non-increasing order of the market value.
- Since the number of players changes from a year to another, we calculate the ranking as the percentage position on the table, i.e., $i/L$ where $i=1,\ldots,L$ is the position of the player in the table, and $L$ is the length of the table, or number of players.

## 3.1. Dividing players based on their birth year and age
Let us first identify the unique birth years


In [16]:
values['birth'].describe()

/var/folders/tw/m4_1qbzn2ybbmcwkchr6k6_c0000gp/T/ipykernel_2774/3604468148.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  values['birth'].describe()


count                    9556
unique                   2595
top       1994-05-27 00:00:00
freq                       22
first     1977-01-02 00:00:00
last      2001-08-16 00:00:00
Name: birth, dtype: object

The years seem to range from 1977 to 2001.

In [17]:
years = [i for i in range(1977,2001,1)]
years

[1977,
 1978,
 1979,
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000]

We are likely to be interested in predicting market values for a limited number of ages. Let us say, from 21 to 28, i.e., not too far after the peak value. 

In [18]:
ages = [a for a in range(20,28,1)]
ages

[20, 21, 22, 23, 24, 25, 26, 27]

In [19]:
players_groups = {}
for y in years:
    for a in ages:
        df = values.loc[(values['birth']> str(y)+'-1-1') 
                        & (values['birth'] < str(y)+'-12-31') 
                        & ( (values['value_at'].dt.year - values['birth'].dt.year) == a),:]
        players_groups[(y,a)] = df

In [20]:
players_groups[(1990,25)]

player_id          player_name         player_role      birth  height  \
9768       133794        Edgar Méndez        Right Winger 1990-01-02    1.87   
9779        73092           Rene Krhin  Defensive Midfield 1990-05-21    1.89   
10313      203043          Pedro Bigas         Centre-Back 1990-05-15    1.81   
10321      223068                 Tana  Attacking Midfield 1990-09-20    1.69   
13754       59344   Asier Illarramendi  Defensive Midfield 1990-03-08    1.79   
...           ...                  ...                 ...        ...     ...   
158151      45184  Grzegorz Krychowiak  Defensive Midfield 1990-01-29    1.87   
158621     183647    Álvaro González         Centre-Back 1990-01-08    1.82   
159548      58426              Douglas          Right-Back 1990-08-06    1.72   
160482      58884                Nacho         Centre-Back 1990-01-18    1.80   
160487      31909           Toni Kroos    Central Midfield 1990-01-04    1.83   

         foot  value league   value_at      nat1 nat2 macro_role  
9768    right   2.25   SPA1 2015-06-30     Spain    -         MF  
9779    right   2.25   SPA1 2015-06-30  Slovenia    -         MF  
10313    left   0.90   SPA1 2015-06-30     Spain    -         DF  
10321    left   0.18   SPA1 2015-06-30     Spain    -         MF  
13754   right  13.50   SPA1 2015-06-30     Spain    -         MF  
...       ...    ...    ...        ...       ...  ...        ...  
158151  right  10.80   SPA1 2015-06-30    Poland    -         MF  
158621  right   2.70   SPA1 2015-06-30     Spain    -         DF  
159548  right   1.80   SPA1 2015-06-30    Brazil    -         DF  
160482  right   5.40   SPA1 2015-06-30     Spain    -         DF  
160487   both  45.00   SPA1 2015-06-30   Germany    -         MF  

[114 rows x 12 columns]

In [21]:
players_groups[(1991,25)]

player_id       player_name         player_role      birth  height  \
5720       127108   Florian Lejeune         Centre-Back 1991-05-20    1.90   
5724       238868      Rubén Peña          Right-Back 1991-07-18    1.70   
9736       129893      Molla Wagué         Centre-Back 1991-02-21    1.91   
10254       96540   Alfredo Ortuño      Centre-Forward 1991-01-21    1.82   
10742      119905        Hugo Mallo          Right-Back 1991-06-22    1.74   
...           ...               ...                 ...        ...     ...   
159764     131505           Rodrigo      Centre-Forward 1991-03-06    1.82   
159774      68645   Martín Montoya          Right-Back 1991-04-14    1.75   
160234     145707            Danilo          Right-Back 1991-07-15    1.84   
160238      88103  James Rodríguez  Attacking Midfield 1991-07-12    1.81   
160258     221316    Lucas Vázquez        Right Winger 1991-07-01    1.73   

         foot  value league   value_at      nat1    nat2 macro_role  
5720    right   0.72   SPA1 2016-06-30    France       -         DF  
5724    right   0.36   SPA1 2016-06-30     Spain       -         DF  
9736    right   1.80   SPA1 2016-06-30      Mali  France         DF  
10254    both   0.90   SPA1 2016-06-30     Spain       -         FW  
10742   right   5.40   SPA1 2016-06-30     Spain       -         DF  
...       ...    ...    ...        ...       ...     ...        ...  
159764   left  10.80   SPA1 2016-06-30     Spain  Brazil         FW  
159774  right   2.70   SPA1 2016-06-30     Spain       -         DF  
160234  right  16.20   SPA1 2016-06-30    Brazil       -         DF  
160238   left  63.00   SPA1 2016-06-30  Colombia   Spain         MF  
160258  right   9.00   SPA1 2016-06-30     Spain       -         MF  

[159 rows x 12 columns]

In [22]:
players_groups[(1987,25)]

player_id       player_name         player_role      birth  height  \
9797        63689         Fran Rico    Central Midfield 1987-08-03    1.78   
13775       37501  Imanol Agirretxe      Centre-Forward 1987-02-24    1.87   
18604       56802  Vincent Muratori           Left-Back 1987-08-03    1.79   
43766       56836     Aaron Meijers           Left-Back 1987-10-28    1.78   
47110       55972      Guram Kashia         Centre-Back 1987-07-04    1.85   
49423       31880    Jordens Peters         Centre-Back 1987-05-03    1.83   
71366       33639    Kevin Mirallas        Right Winger 1987-10-05    1.82   
71976       54170           Ramires    Central Midfield 1987-03-24    1.79   
77216       40204          Joe Hart          Goalkeeper 1987-04-19    1.96   
77231       18935       Samir Nasri  Attacking Midfield 1987-06-26    1.77   
78433       43250    Jan Vertonghen         Centre-Back 1987-04-24    1.89   
78440       19368   Mousa Dembélé    Central Midfield 1987-07-16    1.85   
82994       31835        Mark Noble    Central Midfield 1987-05-08    1.80   
87384       45861    Ryan Shawcross         Centre-Back 1987-10-04    1.91   
87782       37011       Nathan Dyer        Right Winger 1987-11-29    1.65   
89044       69653   Garath McCleary        Right Winger 1987-05-15    1.87   
95545       47085    Nenad Tomović         Centre-Back 1987-08-30    1.84   
97495       39983  Leonardo Bonucci         Centre-Back 1987-05-01    1.90   
104717      29962   Daniele Dessena    Central Midfield 1987-05-10    1.83   
104724      55801         Marco Sau      Centre-Forward 1987-11-03    1.69   
108311      38593      Marek Hamsik    Central Midfield 1987-07-27    1.83   
147759      68815            Felipe         Centre-Back 1987-05-15    1.93   
159561      18944     Gerard Piqué         Centre-Back 1987-02-02    1.92   
159572      28003      Lionel Messi        Right Winger 1987-06-24    1.70   
159587      65278             Pedro        Right Winger 1987-07-28    1.67   
160526      18922     Karim Benzema      Centre-Forward 1987-12-19    1.85   

         foot   value league   value_at         nat1     nat2 macro_role  
9797    right   0.900   SPA1 2012-06-30        Spain        -         MF  
13775   right   2.700   SPA1 2012-06-30        Spain        -         FW  
18604    left   0.900    FR1 2012-06-30       France        -         DF  
43766    left   0.540    NE1 2012-06-30  Netherlands        -         DF  
47110   right   3.600    NE1 2012-06-30      Georgia        -         DF  
49423   right   0.225    NE1 2012-06-30  Netherlands        -         DF  
71366   right   5.850    UK1 2012-06-30      Belgium        -         MF  
71976   right  27.000    UK1 2012-06-30       Brazil        -         MF  
77216   right  21.150    UK1 2012-06-30      England        -         GK  
77231   right  22.500    UK1 2012-06-30       France  Algeria         MF  
78433    left  13.500    UK1 2012-06-30      Belgium        -         DF  
78440    left  11.700    UK1 2012-06-30      Belgium     Mali         MF  
82994   right   6.300    UK1 2012-06-30      England  Ireland         MF  
87384   right   7.200    UK1 2012-06-30      England    Wales         DF  
87782   right   4.050    UK1 2012-06-30      England  Jamaica         MF  
89044   right   0.900    UK1 2012-06-30      Jamaica  England         MF  
95545   right   1.170   ITA1 2012-06-30       Serbia        -         DF  
97495   right  12.150   ITA1 2012-06-30        Italy        -         DF  
104717  right   1.620   ITA1 2012-06-30        Italy        -         MF  
104724  right   1.620   ITA1 2012-06-30        Italy        -         FW  
108311   both  22.500   ITA1 2012-06-30     Slovakia        -         MF  
147759   left   2.700    GE1 2012-06-30       Brazil    Italy         DF  
159561  right  34.200   SPA1 2012-06-30        Spain        -         DF  
159572   left  90.000   SPA1 2012-06-30    Argentina    Spain         MF  
159587   both  25.200   SP

## 3.2. We assign a rank to each record

In [23]:
for y in years:
    for a in ages:
        players_groups[(y,a)]['rank'] = players_groups[(y,a)]['value'].rank(pct=True,ascending=False,method='average',na_option='bottom')


/var/folders/tw/m4_1qbzn2ybbmcwkchr6k6_c0000gp/T/ipykernel_2774/2769782355.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  players_groups[(y,a)]['rank'] = players_groups[(y,a)]['value'].rank(pct=True,ascending=False,method='average',na_option='bottom')


In [24]:
players_groups[(1990,26)]

player_id        player_name         player_role      birth  height  \
5729       153427              Bebé         Left Winger 1990-07-12    1.90   
8221        42920       Fran Mérida    Central Midfield 1990-03-04    1.74   
9721        73092         Rene Krhin  Defensive Midfield 1990-05-21    1.89   
9737        93759   Matthieu Saunier         Centre-Back 1990-02-07    1.81   
10232      223068               Tana  Attacking Midfield 1990-09-20    1.69   
...           ...                ...                 ...        ...     ...   
158377     183647  Álvaro González         Centre-Back 1990-01-08    1.82   
159298      58426            Douglas          Right-Back 1990-08-06    1.72   
159748     227805    Jaume Doménech          Goalkeeper 1990-11-05    1.85   
160231      58884              Nacho         Centre-Back 1990-01-18    1.80   
160253      31909         Toni Kroos    Central Midfield 1990-01-04    1.83   

         foot  value league   value_at      nat1       nat2 macro_role  \
5729    right   1.80   SPA1 2016-06-30  Portugal  CapeVerde         MF   
8221     left   0.54   SPA1 2016-06-30     Spain          -         MF   
9721    right   2.70   SPA1 2016-06-30  Slovenia          -         MF   
9737    right   0.90   SPA1 2016-06-30    France          -         DF   
10232    left   1.35   SPA1 2016-06-30     Spain          -         MF   
...       ...    ...    ...        ...       ...        ...        ...   
158377  right   3.60   SPA1 2016-06-30     Spain          -         DF   
159298  right   0.90   SPA1 2016-06-30    Brazil          -         DF   
159748  right   3.60   SPA1 2016-06-30     Spain          -         GK   
160231  right   4.50   SPA1 2016-06-30     Spain          -         DF   
160253   both  45.00   SPA1 2016-06-30   Germany          -         MF   

            rank  
5729    0.678808  
8221    0.933775  
9721    0.579470  
9737    0.847682  
10232   0.771523  
...          ...  
158377  0.490066  
159298  0.847682  
159748  0.490066  
160231  0.427152  
160253  0.006623  

[151 rows x 13 columns]

# 4. Ratings

In [25]:
ratings = pd.read_csv('../data/ratings.csv',names = ['player_id_lm','player_id','player_name','birth_year','date_rating_18','rating_18','peak_rating_18','minutes_played_18','date_rating_19','rating_19','peak_rating_19','minutes_played_19','date_rating_20','rating_20','peak_rating_20','minutes_played_20'])

In [26]:
ratings.head()

player_id_lm  player_id        player_name  birth_year date_rating_18  \
0         39151      94308    Marko Dmitrovic        1992     30.06.2010   
1         11673     139336     Paulo Oliveira        1992     30.06.2010   
2          6213      87469         Jose Angel        1989     30.06.2007   
3         52516     266795  Gonzalo Escalante        1993     30.06.2011   
4          3184     153427               Bebe        1990     30.06.2008   

   rating_18  peak_rating_18  minutes_played_18 date_rating_19  rating_19  \
0        NaN             NaN                  0     30.06.2011  -0.056574   
1        NaN             NaN                  0     30.06.2011        NaN   
2        NaN             NaN                  0     30.06.2008        NaN   
3        NaN             NaN                  0     30.06.2012        NaN   
4        NaN             NaN                  0     30.06.2009        NaN   

   peak_rating_19  minutes_played_19 date_rating_20  rating_20  \
0          0.0528                 90     30.06.2012  -0.055770   
1             NaN                  0     30.06.2012        NaN   
2             NaN                  0     30.06.2009  -0.023174   
3             NaN                  0     30.06.2013  -0.072950   
4             NaN                  0     30.06.2010        NaN   

   peak_rating_20  minutes_played_20  
0        0.029722                180  
1             NaN                  0  
2        0.051290                992  
3        0.013075                277  
4             NaN                  0

Let us select only the ratings at the age of 20.

In [27]:
ratings_20 = ratings.loc[:,['player_id','player_name','birth_year','date_rating_20','rating_20','peak_rating_20','minutes_played_20']]
ratings_20.head()

player_id        player_name  birth_year date_rating_20  rating_20  \
0      94308    Marko Dmitrovic        1992     30.06.2012  -0.055770   
1     139336     Paulo Oliveira        1992     30.06.2012        NaN   
2      87469         Jose Angel        1989     30.06.2009  -0.023174   
3     266795  Gonzalo Escalante        1993     30.06.2013  -0.072950   
4     153427               Bebe        1990     30.06.2010        NaN   

   peak_rating_20  minutes_played_20  
0        0.029722                180  
1             NaN                  0  
2        0.051290                992  
3        0.013075                277  
4             NaN                  0

From these let us filter out missing values.

In [28]:
ratings_20 = ratings_20.dropna(subset=['rating_20', 'peak_rating_20'])
ratings_20

player_id            player_name  birth_year date_rating_20  rating_20  \
0         94308        Marko Dmitrovic        1992     30.06.2012  -0.055770   
2         87469             Jose Angel        1989     30.06.2009  -0.023174   
3        266795      Gonzalo Escalante        1993     30.06.2013  -0.072950   
5        142027          Pablo Hervias        1993     30.06.2013  -0.096118   
8        138935         Sergio Alvarez        1992     30.06.2012  -0.077713   
...         ...                    ...         ...            ...        ...   
1750      44501                Marcelo        1988     30.06.2008   0.105363   
1751     138927  Daniel Carvajal Ramos        1992     30.06.2012   0.041555   
1753      39381      Gareth Frank Bale        1989     30.06.2009   0.024184   
1754      18922          Karim Benzema        1987     30.06.2007   0.077195   
1755      33648         Fabio Coentrao        1988     30.06.2008   0.049237   

      peak_rating_20  minutes_played_20  
0           0.029722                180  
2           0.051290                992  
3           0.013075                277  
5          -0.011392               1732  
8           0.007705               2003  
...              ...                ...  
1750        0.183862               2844  
1751        0.126091               5650  
1753        0.090404               3629  
1754        0.184693                199  
1755        0.122375                 44  

[918 rows x 7 columns]

In [29]:
ratings_20['birth_year'].describe()

count     918.000000
mean     1990.921569
std         1.747518
min      1987.000000
25%      1990.000000
50%      1991.000000
75%      1992.000000
max      1993.000000
Name: birth_year, dtype: float64

We assign a rank to ratings and minutes played

In [30]:
ratings_20['rank_rating_20'] = ratings_20['rating_20'].rank(pct=True,ascending=False,method='average',na_option='bottom')
ratings_20['rank_peak_rating_20'] = ratings_20['peak_rating_20'].rank(pct=True,ascending=False,method='average',na_option='bottom')
ratings_20['rank_minutes_played_20'] = ratings_20['minutes_played_20'].rank(pct=True,ascending=False,method='average',na_option='bottom')
ratings_20.head()

player_id        player_name  birth_year date_rating_20  rating_20  \
0      94308    Marko Dmitrovic        1992     30.06.2012  -0.055770   
2      87469         Jose Angel        1989     30.06.2009  -0.023174   
3     266795  Gonzalo Escalante        1993     30.06.2013  -0.072950   
5     142027      Pablo Hervias        1993     30.06.2013  -0.096118   
8     138935     Sergio Alvarez        1992     30.06.2012  -0.077713   

   peak_rating_20  minutes_played_20  rank_rating_20  rank_peak_rating_20  \
0        0.029722                180        0.734205             0.745098   
2        0.051290                992        0.537037             0.618736   
3        0.013075                277        0.826797             0.842048   
5       -0.011392               1732        0.925926             0.941176   
8        0.007705               2003        0.857298             0.870370   

   rank_minutes_played_20  
0                0.778322  
2                0.501089  
3                0.720588  
5                0.350763  
8                0.310458

# 5. Regression

We will now set up a regression task where we use the ratings and minutes played at the age of 20 (along with other characteristics of the player) to predict the value ranking at the age of 25. Our regression task is characterized by the response variable (the rank) being bound between 0 and 1. In this case we talk about *censored* regression.
For this reason we use methods that are able to predict the response in its domain.

## 5.1. Data collection

As we have seen, the birth year of the ratings ranges from 1987 to 1993.

In [31]:
years = [i for i in range(1987,1993,1)]
print(years)

[1987, 1988, 1989, 1990, 1991, 1992]


Let us retrieve the appropriate rankings.

In [32]:
rankings = None
for y in years:
    df = players_groups[(y,25)]
    if rankings is None:
        rankings = df
    else:
        rankings = rankings.append(df)
print(rankings)

        player_id        player_name         player_role      birth  height  \
9797        63689          Fran Rico    Central Midfield 1987-08-03    1.78   
13775       37501   Imanol Agirretxe      Centre-Forward 1987-02-24    1.87   
18604       56802   Vincent Muratori           Left-Back 1987-08-03    1.79   
43766       56836      Aaron Meijers           Left-Back 1987-10-28    1.78   
47110       55972       Guram Kashia         Centre-Back 1987-07-04    1.85   
...           ...                ...                 ...        ...     ...   
159350      85370      Sergi Roberto          Right-Back 1992-02-07    1.78   
159357      80444  Philippe Coutinho         Left Winger 1992-06-12    1.72   
159789     131102     Jeison Murillo         Centre-Back 1992-05-27    1.82   
160282     138927    Daniel Carvajal          Right-Back 1992-01-11    1.73   
160287      85288               Isco  Attacking Midfield 1992-04-21    1.76   

         foot  value league   value_at         nat1

We have 918 ratings for the age of 20 and 616 rankings. We should now find the players for which we have both a rating and a ranking. First, let us select only the columns which for now necessary or we want to use in regression.

In [33]:
rankings = rankings.loc[:,['player_id','player_name','player_role','macro_role','birth','height','foot','rank']]

In [34]:
data = ratings_20.merge(rankings,on='player_id', how='inner')

In [35]:
data

player_id          player_name_x  birth_year date_rating_20  rating_20  \
0        94308        Marko Dmitrovic        1992     30.06.2012  -0.055770   
1       127108        Florian Lejeune        1991     30.06.2011  -0.067635   
2       129679        Cheick Doukoure        1992     30.06.2012  -0.016803   
3        73092             Rene Krhin        1990     30.06.2010   0.056437   
4       133794           Edgar Mendez        1991     30.06.2011  -0.099999   
..         ...                    ...         ...            ...        ...   
373      44501                Marcelo        1988     30.06.2008   0.105363   
374     138927  Daniel Carvajal Ramos        1992     30.06.2012   0.041555   
375      39381      Gareth Frank Bale        1989     30.06.2009   0.024184   
376      18922          Karim Benzema        1987     30.06.2007   0.077195   
377      33648         Fabio Coentrao        1988     30.06.2008   0.049237   

     peak_rating_20  minutes_played_20  rank_rating_20  rank_peak_rating_20  \
0          0.029722                180        0.734205             0.745098   
1          0.018432               3689        0.799564             0.815904   
2          0.087269                153        0.483660             0.326797   
3          0.130754                 74        0.067538             0.090414   
4         -0.015180               1163        0.935730             0.955338   
..              ...                ...             ...                  ...   
373        0.183862               2844        0.011983             0.014161   
374        0.126091               5650        0.124183             0.106754   
375        0.090404               3629        0.210240             0.307190   
376        0.184693                199        0.043573             0.013072   
377        0.122375                 44        0.088235             0.132898   

     rank_minutes_played_20     player_name_y         player_role macro_role  \
0                  0.778322   Marko Dmitrovic          Goalkeeper         GK   
1                  0.119826   Florian Lejeune         Centre-Back         DF   
2                  0.793573  Cheick Doukouré  Defensive Midfield         MF   
3                  0.874728        Rene Krhin  Defensive Midfield         MF   
4                  0.456427     Edgar Méndez        Right Winger         MF   
..                      ...               ...                 ...        ...   
373                0.184096           Marcelo           Left-Back         DF   
374                0.039216   Daniel Carvajal          Right-Back         DF   
375                0.125272       Gareth Bale        Right Winger         MF   
376                0.757081     Karim Benzema      Centre-Forward         FW   
377                0.912309  Fábio Coentrão           Left-Back         DF   

         birth  height   foot      rank  
0   1992-01-24    1.94   left  0.827830  
1   1991-05-20    1.90  right  0.902516  
2   1992-09-11    1.80  right  0.780660  
3   1990-05-21    1.89  right  0.649123  
4   1990-01-02    1.87  right  0.649123  
..         ...     ...    ...       ...  
373 1988-05-12    1.74   left  0.179487  
374 1992-01-11    1.73  right  0.070755  
375 1989-07-16    1.85   left  0.015152  
376 1987-12-19    1.85   both  0.115385  
377 1988-03-11    1.79   left  0.243590  

[378 rows x 17 columns]

It looks like we have ratings and rankings for 378 players. This is the data we can use to do regression. 

Let us add the value at the age of 20.

In [36]:
rankings20 = None
for y in years:
    df = players_groups[(y,20)]
    if rankings20 is None:
        rankings20 = df
    else:
        rankings20 = rankings20.append(df)
print(rankings20)

        player_id          player_name         player_role      birth  height  \
77052       40204             Joe Hart          Goalkeeper 1987-04-19    1.96   
82798       31835           Mark Noble    Central Midfield 1987-05-08    1.80   
156469      41319       Patrick Fabian         Centre-Back 1987-10-11    1.94   
159409      28003         Lionel Messi        Right Winger 1987-06-24    1.70   
149121      35208     Marcel Schmelzer           Left-Back 1988-01-22    1.80   
149529      42203  Fabian Lustenberger         Centre-Back 1988-05-02    1.80   
153090      39020    Benedikt Höwedes         Centre-Back 1988-02-29    1.88   
160412      44501              Marcelo           Left-Back 1988-05-12    1.74   
70544       50362         Fabian Delph    Central Midfield 1989-11-21    1.74   
79547       33713         Theo Walcott        Right Winger 1989-03-16    1.75   
82827       61592        James Tomkins         Centre-Back 1989-03-29    1.92   
149080      29993          S

## 5.2. Random Forests version 1

In this version we use the following regressors: rating, peak_rating, and minutes played at the age of 20.
We start by importing the necessary packages.

In [37]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, GridSearchCV

### 5.2.1 Data selection

In [38]:
Y = data['rank']
Y

0      0.827830
1      0.902516
2      0.780660
3      0.649123
4      0.649123
         ...   
373    0.179487
374    0.070755
375    0.015152
376    0.115385
377    0.243590
Name: rank, Length: 378, dtype: float64

In [39]:
X1 = data.loc[:,['rating_20','peak_rating_20','minutes_played_20']]
X1

rating_20  peak_rating_20  minutes_played_20
0    -0.055770        0.029722                180
1    -0.067635        0.018432               3689
2    -0.016803        0.087269                153
3     0.056437        0.130754                 74
4    -0.099999       -0.015180               1163
..         ...             ...                ...
373   0.105363        0.183862               2844
374   0.041555        0.126091               5650
375   0.024184        0.090404               3629
376   0.077195        0.184693                199
377   0.049237        0.122375                 44

[378 rows x 3 columns]

In [40]:
X1_train, X1_test, Y1_train, Y1_test = train_test_split(X1,Y)

### 5.2.2 Grid Search for optimal parameters setup

We specify the parameters of the RF regressor that we want to optimize using Grid Search cross-validation.

In [41]:
RandomForestRegressor().get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'squared_error',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [42]:
parameters = [
    {"max_depth": [2, 4, 6],'min_samples_split': [2,4,6], 'min_samples_leaf':[1,2,3]}
]

We perform grid search

In [43]:
gs = GridSearchCV(RandomForestRegressor(), parameters, scoring='neg_root_mean_squared_error')
gs.fit(X1_train, Y1_train)

GridSearchCV(estimator=RandomForestRegressor(),
             param_grid=[{'max_depth': [2, 4, 6], 'min_samples_leaf': [1, 2, 3],
                          'min_samples_split': [2, 4, 6]}],
             scoring='neg_root_mean_squared_error')

In [44]:
gs.best_params_

{'max_depth': 2, 'min_samples_leaf': 3, 'min_samples_split': 4}

In [45]:
gs.cv_results_["mean_test_score"]

array([-0.23744623, -0.23758821, -0.23618262, -0.23654443, -0.23659766,
       -0.23620879, -0.23594818, -0.23516767, -0.2365195 , -0.24040152,
       -0.24062346, -0.24195046, -0.23933686, -0.24041427, -0.23908688,
       -0.23944653, -0.23989576, -0.2416198 , -0.24634529, -0.24406847,
       -0.24404235, -0.24473081, -0.24469738, -0.24452769, -0.24455609,
       -0.243532  , -0.24611187])

## 5.3 Random Forests version 2

In this version we add the role as a predictor and repeat the entire experiment.

### 5.3.1 We select the data

In [46]:
X2 = data.loc[:,['rating_20','peak_rating_20','minutes_played_20','player_role']]
X2

rating_20  peak_rating_20  minutes_played_20         player_role
0    -0.055770        0.029722                180          Goalkeeper
1    -0.067635        0.018432               3689         Centre-Back
2    -0.016803        0.087269                153  Defensive Midfield
3     0.056437        0.130754                 74  Defensive Midfield
4    -0.099999       -0.015180               1163        Right Winger
..         ...             ...                ...                 ...
373   0.105363        0.183862               2844           Left-Back
374   0.041555        0.126091               5650          Right-Back
375   0.024184        0.090404               3629        Right Winger
376   0.077195        0.184693                199      Centre-Forward
377   0.049237        0.122375                 44           Left-Back

[378 rows x 4 columns]

We use one-hot encoding for the role.

In [47]:
one_hot = pd.get_dummies(X2['player_role'],drop_first=True)
one_hot

Central Midfield  Centre-Back  Centre-Forward  Defender  \
0                   0            0               0         0   
1                   0            1               0         0   
2                   0            0               0         0   
3                   0            0               0         0   
4                   0            0               0         0   
..                ...          ...             ...       ...   
373                 0            0               0         0   
374                 0            0               0         0   
375                 0            0               0         0   
376                 0            0               1         0   
377                 0            0               0         0   

     Defensive Midfield  Forward  Goalkeeper  Left Midfield  Left Winger  \
0                     0        0           1              0            0   
1                     0        0           0              0            0   
2                     1        0           0              0            0   
3                     1        0           0              0            0   
4                     0        0           0              0            0   
..                  ...      ...         ...            ...          ...   
373                   0        0           0              0            0   
374                   0        0           0              0            0   
375                   0        0           0              0            0   
376                   0        0           0              0            0   
377                   0        0           0              0            0   

     Left-Back  Midfielder  Right Midfield  Right Winger  Right-Back  \
0            0           0               0             0           0   
1            0           0               0             0           0   
2            0           0               0             0           0   
3            0           0               0             0           0   
4            0           0               0             1           0   
..         ...         ...             ...           ...         ...   
373          1           0               0             0           0   
374          0           0               0             0           1   
375          0           0               0             1           0   
376          0           0               0             0           0   
377          1           0               0             0           0   

     Second Striker  Sweeper  
0                 0        0  
1                 0        0  
2                 0        0  
3                 0        0  
4                 0        0  
..              ...      ...  
373               0        0  
374               0        0  
375               0        0  
376               0        0  
377               0        0  

[378 rows x 16 columns]

In [48]:
X2 = X2.drop('player_role',axis = 1)
X2 = X2.join(one_hot)
X2.head()

rating_20  peak_rating_20  minutes_played_20  Central Midfield  \
0  -0.055770        0.029722                180                 0   
1  -0.067635        0.018432               3689                 0   
2  -0.016803        0.087269                153                 0   
3   0.056437        0.130754                 74                 0   
4  -0.099999       -0.015180               1163                 0   

   Centre-Back  Centre-Forward  Defender  Defensive Midfield  Forward  \
0            0               0         0                   0        0   
1            1               0         0                   0        0   
2            0               0         0                   1        0   
3            0               0         0                   1        0   
4            0               0         0                   0        0   

   Goalkeeper  Left Midfield  Left Winger  Left-Back  Midfielder  \
0           1              0            0          0           0   
1           0              0            0          0           0   
2           0              0            0          0           0   
3           0              0            0          0           0   
4           0              0            0          0           0   

   Right Midfield  Right Winger  Right-Back  Second Striker  Sweeper  
0               0             0           0               0        0  
1               0             0           0               0        0  
2               0             0           0               0        0  
3               0             0           0               0        0  
4               0             1           0               0        0

In [49]:
X2_train, X2_test, Y2_train, Y2_test = train_test_split(X2,Y)

### 5.3.2. We use grid search to find the best parameters

In [50]:
parameters = [
    {"max_depth": [2, 4, 6],'min_samples_split': [2,4,6], 'min_samples_leaf':[1,2,3]}
]
gs = GridSearchCV(RandomForestRegressor(), parameters, scoring='neg_root_mean_squared_error')
gs.fit(X2_train, Y2_train)
gs.best_params_

{'max_depth': 2, 'min_samples_leaf': 2, 'min_samples_split': 4}

In [51]:
gs.cv_results_["mean_test_score"]

array([-0.23882161, -0.23963454, -0.23893617, -0.23913303, -0.23745684,
       -0.23906695, -0.23808983, -0.2376854 , -0.23872365, -0.24103522,
       -0.23946869, -0.24015036, -0.2407194 , -0.24147904, -0.23993208,
       -0.24228209, -0.24117151, -0.23938541, -0.24250734, -0.2425307 ,
       -0.24319596, -0.24446605, -0.2449947 , -0.24361918, -0.2438534 ,
       -0.24257022, -0.24419469])

We can notice that the RMSE is comparable to that of the RF version without the role as a predictor.

## 5.4. Random Forests version 3

In this version we standardize the values (except the one-hot encoding of the role).

### 5.4.1. We gather the data

In [52]:
from sklearn.preprocessing import StandardScaler
X3 = X2
X3[['rating_20', 'peak_rating_20','minutes_played_20']] = StandardScaler().fit_transform(X3[['rating_20', 'peak_rating_20','minutes_played_20']])
X3.head()

rating_20  peak_rating_20  minutes_played_20  Central Midfield  \
0  -0.785149       -0.844500          -0.886444                 0   
1  -1.013924       -1.069490           1.065738                 0   
2  -0.033824        0.302281          -0.901465                 0   
3   1.378319        1.168840          -0.945416                 0   
4  -1.637936       -1.739289          -0.339566                 0   

   Centre-Back  Centre-Forward  Defender  Defensive Midfield  Forward  \
0            0               0         0                   0        0   
1            1               0         0                   0        0   
2            0               0         0                   1        0   
3            0               0         0                   1        0   
4            0               0         0                   0        0   

   Goalkeeper  Left Midfield  Left Winger  Left-Back  Midfielder  \
0           1              0            0          0           0   
1           0              0            0          0           0   
2           0              0            0          0           0   
3           0              0            0          0           0   
4           0              0            0          0           0   

   Right Midfield  Right Winger  Right-Back  Second Striker  Sweeper  
0               0             0           0               0        0  
1               0             0           0               0        0  
2               0             0           0               0        0  
3               0             0           0               0        0  
4               0             1           0               0        0

### 5.4.2. Grid Search for best parameters

In [53]:
X3_train, X3_test, Y3_train, Y3_test = train_test_split(X3,Y)

In [54]:
parameters = [
    {"max_depth": [2, 4, 6],'min_samples_split': [2,4,6], 'min_samples_leaf':[1,2,3]}
]
gs = GridSearchCV(RandomForestRegressor(), parameters, scoring='neg_root_mean_squared_error')
gs.fit(X3_train, Y3_train)
gs.best_params_

{'max_depth': 2, 'min_samples_leaf': 2, 'min_samples_split': 2}

In [55]:
gs.cv_results_["mean_test_score"]

array([-0.2387678 , -0.23917566, -0.23889019, -0.23825387, -0.2395289 ,
       -0.23849462, -0.2391706 , -0.23970586, -0.23842669, -0.2429239 ,
       -0.24285347, -0.24295008, -0.24255547, -0.24296708, -0.24347592,
       -0.24307583, -0.24270624, -0.24438394, -0.24537783, -0.24452843,
       -0.24679624, -0.24465956, -0.24486222, -0.2464496 , -0.2447791 ,
       -0.24760003, -0.24603357])

Also in this case we do not obtain a significant improvement in the prediction accuracy of the model.

### 5.4.3 Creation of retained RF predictor with optimal parameters

Let us now create a Random Forest predictor with the parameters obtained in the grid search

In [56]:
X3_train, X3_test, Y3_train, Y3_test = train_test_split(X3,Y)
rf3 = RandomForestRegressor(max_depth=4,min_samples_leaf=1,min_samples_split=6)
rf3.fit(X3_train,Y3_train)

RandomForestRegressor(max_depth=4, min_samples_split=6)

In [57]:
Y3_pred = rf3.predict(X3_test)

In [58]:
rf3.score(X3_test,Y3_test)

0.251947327454795

In [59]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(Y3_test, Y3_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(Y3_test, Y3_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(Y3_test, Y3_pred)))
print('Mean Absolute Percentage Error:', metrics.mean_absolute_percentage_error(Y3_test,Y3_pred))

Mean Absolute Error: 0.20344061816033018
Mean Squared Error: 0.057185875182502684
Root Mean Squared Error: 0.23913568362438653
Mean Absolute Percentage Error: 1.7728763388484956


As we can see we have a MAE of 0.19 which means that on average we are 20% off the right value ranking.

## 5.5. Random Forest version 4
In this version we use the ranks of ratings and minutes played as predictors.

### 5.5.1. Data selection

In [60]:
Y = data['rank']
Y

0      0.827830
1      0.902516
2      0.780660
3      0.649123
4      0.649123
         ...   
373    0.179487
374    0.070755
375    0.015152
376    0.115385
377    0.243590
Name: rank, Length: 378, dtype: float64

In [61]:
X4 = data.loc[:,['rank_rating_20','rank_peak_rating_20','rank_minutes_played_20']]

In [62]:
one_hot = pd.get_dummies(data['macro_role'])
X4 = X4.join(one_hot)
X4.head()

rank_rating_20  rank_peak_rating_20  rank_minutes_played_20  DF  FW  GK  MF
0        0.734205             0.745098                0.778322   0   0   1   0
1        0.799564             0.815904                0.119826   1   0   0   0
2        0.483660             0.326797                0.793573   0   0   0   1
3        0.067538             0.090414                0.874728   0   0   0   1
4        0.935730             0.955338                0.456427   0   0   0   1

### 5.5.2. Grid search for best parameters

In [63]:
X4_train, X4_test, Y4_train, Y4_test = train_test_split(X4,Y)
parameters = [
    {"max_depth": [2, 4, 6],'min_samples_split': [2,4,6], 'min_samples_leaf':[1,2,3]}
]
gs4 = GridSearchCV(RandomForestRegressor(), parameters, scoring='neg_root_mean_squared_error')
gs4.fit(X4_train, Y4_train)
gs4.best_params_

{'max_depth': 2, 'min_samples_leaf': 2, 'min_samples_split': 4}

In [64]:
gs4.cv_results_["mean_test_score"]

array([-0.25103918, -0.25129436, -0.25110947, -0.25048609, -0.24992568,
       -0.25121972, -0.25127628, -0.25119734, -0.25251678, -0.25333576,
       -0.25363528, -0.25516165, -0.25422998, -0.25431726, -0.25507616,
       -0.25622994, -0.2545958 , -0.2533421 , -0.25712211, -0.25516758,
       -0.2571348 , -0.25695571, -0.25723795, -0.25900071, -0.2580292 ,
       -0.25772991, -0.25742112])

## 5.5. Random Forest version 5
We add the value rank at the age of 20 as a predictor.

In [65]:
X5 = data.loc[:,['rank','player_id','rank_rating_20','rank_peak_rating_20','rank_minutes_played_20']]
X5 = X5.merge(rankings20.loc[:,['player_id','rank']],on='player_id', how='inner')
X5

rank_x  player_id  rank_rating_20  rank_peak_rating_20  \
0   0.153302      74230        0.217865             0.174292   
1   0.450472      99352        0.680828             0.530501   
2   0.364035      73250        0.676471             0.622004   
3   0.990566     123160        0.433551             0.417211   
4   0.014151     102558        0.410675             0.252723   
5   0.051887      77100        0.057734             0.034858   
6   0.051887      77100        0.057734             0.034858   
7   0.115566     161869        0.749455             0.683007   
8   0.827830     126681        0.370370             0.324619   
9   0.037736      93720        0.194989             0.185185   
10  0.588050     127181        0.494553             0.502179   
11  0.945755     195359        0.613290             0.581699   
12  0.060606      33713        0.002179             0.003268   
13  0.092105      95424        0.506536             0.600218   
14  0.307692      40204        0.156863             0.163399   
15  0.172170     117996        0.019608             0.015251   
16  0.210526      50174        0.614379             0.679739   
17  0.141509      72476        0.220044             0.201525   
18  0.400943     126321        0.500000             0.491285   
19  0.210692      74869        0.399782             0.387800   
20  0.030702      50057        0.064270             0.049020   
21  0.014151      80444        0.181917             0.115468   
22  0.014151      80444        0.181917             0.115468   
23  0.697368     121477        0.635076             0.717865   
24  0.469697      61592        0.318083             0.516340   
25  0.500000      31835        0.152505             0.151416   
26  0.364035     108484        0.290850             0.242919   
27  0.129717      86784        0.247277             0.159041   
28  0.433962      82350        0.565359             0.503268   
29  0.429245      94201        0.361656             0.291939   
30  0.159091      29993        0.572985             0.729847   
31  0.717949      42203        0.355120             0.416122   
32  0.051887      59016        0.040305             0.026144   
33  0.030303      58358        0.016340             0.028322   
34  0.090909      30059        0.449891             0.636166   
35  0.243590      39020        0.071895             0.104575   
36  0.325758      96718        0.715686             0.864924   
37  0.153302      54235        0.098039             0.047930   
38  0.038462      28003        0.003268             0.002179   
39  0.179487      44501        0.011983             0.014161   

    rank_minutes_played_20    rank_y  
0                 0.107843  0.475000  
1                 0.196078  0.675000  
2                 0.203704  0.571429  
3                 0.772331  0.666667  
4                 0.102397  0.250000  
5                 0.877996  0.125000  
6                 0.877996  0.125000  
7                 0.141612  0.550000  
8                 0.722222  0.750000  
9                 0.751089  0.250000  
10                0.901961  0.833333  
11                0.485839  0.950000  
12                0.044662  0.125000  
13                0.172113  0.285714  
14                0.855664  0.500000  
15                0.017429  0.125000  
16                0.459150  0.714286  
17                0.114379  0.350000  
18                0.377996  0.675000  
19                0.454793  0.500000  
20                0.258170  0.142857  
21                0.216776  0.400000  
22                0.216776  0.400000  
23                0.130719  0.857143  
24                0.298475  0.250000  
25                0.453159  0.750000  
26                0.633987  0.428571  
27                0.049020  0.300000  
28                0.285403  0.250000  
29                0.921569  0.900000  
30                0.357298  0.500000  
31                0.412854  0.500000  
32                0.043573  0.200000  
33                0.906318  0.625000  
34              

In [66]:
Y5 = X5['rank_x']
Y5

0     0.153302
1     0.450472
2     0.364035
3     0.990566
4     0.014151
5     0.051887
6     0.051887
7     0.115566
8     0.827830
9     0.037736
10    0.588050
11    0.945755
12    0.060606
13    0.092105
14    0.307692
15    0.172170
16    0.210526
17    0.141509
18    0.400943
19    0.210692
20    0.030702
21    0.014151
22    0.014151
23    0.697368
24    0.469697
25    0.500000
26    0.364035
27    0.129717
28    0.433962
29    0.429245
30    0.159091
31    0.717949
32    0.051887
33    0.030303
34    0.090909
35    0.243590
36    0.325758
37    0.153302
38    0.038462
39    0.179487
Name: rank_x, dtype: float64

In [67]:
X5 = X5.loc[:,['rank_rating_20','rank_peak_rating_20','rank_minutes_played_20','rank_y']]
X5

rank_rating_20  rank_peak_rating_20  rank_minutes_played_20    rank_y
0         0.217865             0.174292                0.107843  0.475000
1         0.680828             0.530501                0.196078  0.675000
2         0.676471             0.622004                0.203704  0.571429
3         0.433551             0.417211                0.772331  0.666667
4         0.410675             0.252723                0.102397  0.250000
5         0.057734             0.034858                0.877996  0.125000
6         0.057734             0.034858                0.877996  0.125000
7         0.749455             0.683007                0.141612  0.550000
8         0.370370             0.324619                0.722222  0.750000
9         0.194989             0.185185                0.751089  0.250000
10        0.494553             0.502179                0.901961  0.833333
11        0.613290             0.581699                0.485839  0.950000
12        0.002179             0.003268                0.044662  0.125000
13        0.506536             0.600218                0.172113  0.285714
14        0.156863             0.163399                0.855664  0.500000
15        0.019608             0.015251                0.017429  0.125000
16        0.614379             0.679739                0.459150  0.714286
17        0.220044             0.201525                0.114379  0.350000
18        0.500000             0.491285                0.377996  0.675000
19        0.399782             0.387800                0.454793  0.500000
20        0.064270             0.049020                0.258170  0.142857
21        0.181917             0.115468                0.216776  0.400000
22        0.181917             0.115468                0.216776  0.400000
23        0.635076             0.717865                0.130719  0.857143
24        0.318083             0.516340                0.298475  0.250000
25        0.152505             0.151416                0.453159  0.750000
26        0.290850             0.242919                0.633987  0.428571
27        0.247277             0.159041                0.049020  0.300000
28        0.565359             0.503268                0.285403  0.250000
29        0.361656             0.291939                0.921569  0.900000
30        0.572985             0.729847                0.357298  0.500000
31        0.355120             0.416122                0.412854  0.500000
32        0.040305             0.026144                0.043573  0.200000
33        0.016340             0.028322                0.906318  0.625000
34        0.449891             0.636166                0.474946  0.375000
35        0.071895             0.104575                0.681917  0.750000
36        0.715686             0.864924                0.540305  0.750000
37        0.098039             0.047930                0.006536  0.050000
38        0.003268             0.002179                0.091503  0.250000
39        0.011983             0.014161                0.184096  0.250000

In [68]:
X5_train, X5_test, Y5_train, Y5_test = train_test_split(X5,Y5)
parameters = [
    {"max_depth": [2, 4, 6],'min_samples_split': [2,4,6], 'min_samples_leaf':[1,2,3]}
]
gs5 = GridSearchCV(RandomForestRegressor(), parameters, scoring='neg_root_mean_squared_error')
gs5.fit(X5_train, Y5_train)
gs5.best_params_

{'max_depth': 4, 'min_samples_leaf': 3, 'min_samples_split': 6}

In [69]:
gs5.cv_results_["mean_test_score"]

array([-0.21674092, -0.21889115, -0.21718123, -0.20957984, -0.20879216,
       -0.20901694, -0.20246017, -0.20229592, -0.20589092, -0.21612465,
       -0.21919303, -0.21235947, -0.20637355, -0.20268012, -0.20865899,
       -0.20250038, -0.20623688, -0.20195065, -0.21560886, -0.21748491,
       -0.21874528, -0.20224326, -0.20652914, -0.21248397, -0.2058858 ,
       -0.20415633, -0.20572035])